azurerm/resource_arm_azure_firewall_network_rule_collection.go

```diff
@@ -0,0 +1,342 @@
+							Required: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Set:      schema.HashString,
+						},
+						"protocols": {
+							Type:     schema.TypeSet,
+							Required: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+								ValidateFunc: validation.StringInSlice([]string{
+									string(network.Any),
+									string(network.ICMP),
+									string(network.TCP),
+									string(network.UDP),
+								}, true),
+							},
+							DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+							Set:              schema.HashString,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceArmAzureFirewallNetworkRuleCollectionCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).azureFirewallsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	resourceGroup := d.Get("resource_group_name").(string)
+	name := d.Get("name").(string)
+	firewallName := d.Get("azure_firewall_name").(string)
+
+	azureRMLockByName(firewallName, azureFirewallResourceName)
+	defer azureRMUnlockByName(firewallName, azureFirewallResourceName)
+
+	firewall, err := client.Get(ctx, resourceGroup, firewallName)
+	if err != nil {
+		return fmt.Errorf("Error retrieving Azure Firewall %q (Resource Group %q): %+v", firewallName, resourceGroup, err)
+	}
+
+	ipConfigs := fixArmAzureFirewallIPConfiguration(&firewall)
+	firewall.AzureFirewallPropertiesFormat.IPConfigurations = &ipConfigs
+
+	newFwRuleCol := expandArmAzureFirewallNetworkRuleCollection(d)
+	ruleCollections := append(*firewall.AzureFirewallPropertiesFormat.NetworkRuleCollections, newFwRuleCol)
+	existingCollection, index, exists := findArmAzureFirewallNetworkRuleCollectionByName(&firewall, name)
+	if exists {
+		if name == *existingCollection.Name {
+			ruleCollections = append(ruleCollections[:index], ruleCollections[index+1:]...)
+		}
+	}
+	firewall.AzureFirewallPropertiesFormat.NetworkRuleCollections = &ruleCollections
+
+	future, err := client.CreateOrUpdate(ctx, resourceGroup, firewallName, firewall)
+	if err != nil {
+		return fmt.Errorf("Error creating/updating Azure Firewall %q (Resource Group %q): %+v", firewallName, resourceGroup, err)
+	}
+
+	err = future.WaitForCompletion(ctx, client.Client)
+	if err != nil {
+		return fmt.Errorf("Error waiting for creation/update of Azure Firewall %q (Resource Group %q): %+v", firewallName, resourceGroup, err)
+	}
+
+	read, err := client.Get(ctx, resourceGroup, firewallName)
+	if err != nil {
+		return fmt.Errorf("Error retrieving Azure Firewall %q (Resource Group %q): %+v", firewallName, resourceGroup, err)
+	}
+
+	if read.ID == nil {
+		return fmt.Errorf("Cannot read Azure Firewall %q (Resource Group %q) ID", firewallName, resourceGroup)
+	}
+
+	var collectionID string
+	for _, collection := range *read.AzureFirewallPropertiesFormat.NetworkRuleCollections {
+		if *collection.Name == name {
+			collectionID = *collection.ID
+		}
+	}
+	d.SetId(collectionID)
```

Comment:
so the ID needs to contain all the components required to import it, as such we'll need to be a little creative here. In general for nested resources we follow the pattern `{parentId}/somename/{name}` - so in this case I'd suggest we use `fmt.Sprintf("%s/networkRuleCollection/%s", *read.ID, *collection.Name)` which will allow us to parse it out below as `id.Path["networkRuleCollection"]`